contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,493 @@
+2. Add a validation step to the CREATE and UPDATE REST handlers of the apiextensions-apiserver.
+ 
+As an optional follow-up, we can implement the OpenAPI part and with that enable client-side validation:
+
+3. Export the JSON-Schema via a dynamically served OpenAPI spec.
+
+## Appendix
+
+### Expressiveness of JSON-Schema
+
+The following example properties cannot be expressed using JSON-Schema:
+1. “In a PodSpec, for each `spec.Containers[*].volumeMounts[*].Name` there must be a `spec.Volumes[*].Name`”
+2. “The volume names in `PodSpec.Volumes` are unique” (`uniqueItems` only compares the complete objects, it cannot compare by key)
+ 
+Different versions within one CRD with a custom version field (i.e. not the one in apiVersion) **can** be expressed:
+
+```json
+{
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "title": "child_schema",
+    "type": "object",
+    "anyOf": [
+        {
+            "properties": {
+                "version": {
+                    "type": "string",
+                    "pattern": "^a$"
+                },
+                "spec": {
+                    "type": "object",
+                    "properties": {
+                        "foo": {}
+                    },
+                    "additionalProperties": false,
+                },
+            }
+        },
+        {
+            "properties": {
+                "version": {
+                    "type": "string",
+                    "pattern": "^b$"
+                },
+                "spec": {
+                    "type": "object",
+                    "properties": {
+                        "bar": {}
+                    },
+                    "additionalProperties": false,
+                },
+            }
+        }
+    ],
+}
+```
+
+This validates: 
+* `{"version": "a",  "spec": {"foo": 42}}`
+* `{"version": "b",  "spec": {"bar": 42}}`
+
+but not:
+* `{"version": "a",  "spec": {"bar": 42}}`.
+ 
+Note: this is a workaround while we do not support multiple versions and conversion for custom resources.
+
+### JSON-Schema Validation Runtime Complexity
+
+Following “JSON: data model, query languages and schema specification<sup id="f4">[4](#footnote4)</sup>” and “Formal Specification, Expressiveness and Complexity analysis for JSON Schema<sup id="f5">[5](#footnote5)</sup>”, JSON-Schema validation
+* without the uniqueItems operator and 
+* without recursion for the $ref operator
+has linear runtime in the size of the JSON input and the size of the schema (Th. 1 plus Prop. 7).
+ 
+If we allow uniqueItems, the runtime complexity becomes quadratic in the size of the JSON input. Hence, we might want to consider forbidding the uniqueItems operator in order to avoid DDoS attacks, at least if the schema definitions of CRDs cannot be trusted.
+
+The CRD JSON-Schema will be validated to have neither recursion, nor `uniqueItems=true` being set.
+
+### Alternatives
```

Comment:
I think many people will see imperative validation (custom code inserted in the request path somehow) as the main alternative to this proposal. You mentioned above they are not mutually exclusive, but it would be good to document here why we think it makes sense to put our effort in creating the declarative version first.